lib/vagrant-service-manager/binary_handler_factory.rb

```diff
@@ -0,0 +1,15 @@
+require_relative 'binary_handlers/docker_binary_handler'
+require_relative 'binary_handlers/openshift_binary_handler'
+
+module VagrantPlugins
+  module ServiceManager
+    class BinaryHandlerFactory
+      def self.build(machine, env, options = {})
+        case options[:type]
+        when :docker then DockerBinaryHandler.new(machine, env, options)
```

Comment:
So your Docker and OpenShift handlers still contain if/else statements for handling ADB vs CDK. Here is the perfect location to handle this and get rid of all this conditional logic. Why not check the box version here as well and load the appropriate handler. `DockerBinaryHandler` becomes ADBDockerBinaryHandler` and the same for Openshift. then you create CDK stub handlers. If in the end the CDK handlers look the "same" as the ADB ones, one can create base classes for these and have only the differences in the respective ADB/CDK subclasses.